### PR TITLE
feat!: initial version of template for templates

### PR DESCRIPTION
**Pull request by derberg**
**Description**

- initial version of template for templates

Let me know if I should rather split it into multiple commits, if this is actually going to help, or better just add some more description here?

This is the result you get when using this template with 
```bash
ag https://bit.ly/asyncapi-dummy https://github.com/derberg/template-for-generator-templates-1/#init -o output
open output/index.html
#or
ag https://bit.ly/asyncapi-dummy https://github.com/derberg/template-for-generator-templates-1/#init -o output -p pdf=true
open output/index.pdf
```

Screenshot 2020-10-14 at 18 30 51
Screenshot 2020-10-14 at 18 31 42

Resolves https://github.com/asyncapi/generator/issues/402

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,617 @@
...
+```
+{% from "partials/listChannels.html" import listChannels %}
+...
+{{ listChannels(asyncapi.channels(), 'subscribe') }}
+{{ listChannels(asyncapi.channels(), 'publish') }}
+```
+
+### File Templates
+
+The Generator has a feature called `file templates` that allows you to create a template file with special that has `$$` markers, like `$$schema$$`. There are multiple different file templates available. In this Template, during generation, `$$schema$$` is replaced with a schema name, and Template gets the following variables in the context:
```
Comment (Tenischev) [inline README.md:426]: Maybe a link to https://github.com/asyncapi/generator/blob/master/docs/authoring.md#file-templates?

File: README.md [hunk 2]
```diff
@@ -0,0 +1,617 @@
...
+  ```
+- Generator is a library that also has a CLI. Parameters passed to the Template using CLI are always of String type. This means that even if your parameter accepts Boolean values like true/false, you still need to check them like a String value
+  ```js
+  const parameters = generator.templateParams;
+  if (parameters['pdf'] === 'true') await page.pdf({ format: 'A4', path: `${targetDir}/index.pdf` });
+  ```
+
+#### Official AsyncAPI Hooks
+
+Hooks are reusable between templates. AsyncAPI Initiative provides a library of hooks. You can also create such a library for your templates. You add such a library to `dependencies` in the `package.json` file and configure in `generator.hooks` section like:
```
Comment (Tenischev) [inline README.md:209]: Maybe a link to https://github.com/asyncapi/generator/blob/master/docs/authoring.md#hooks?

File: partials/diagramContent.html [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    <h2>Schemas</h2>
+    <hr>
+    You have the following schemas named in components section:
+    <div class="container mx-auto px-8">
+        <ul class="list-disc">
+           {# Usage of build-in "dictsort" Nunjucks filter for sorting items alphabetically 
+            and how it works in a for loop #}
+            {% for schema in asyncapi.components().schemas() | dictsort %}
+                {# Declare variables for values that you use more than once, just like you do with regular code #}
+                {% set schemaName = schema[0] %}
```
[the remark this answers is not in the document]
Comment (derberg) [reply partials/diagramContent.html:17]: I added an explanation why in this case I do it this way

**Review by Tenischev — APPROVED** [review]

**Comment by derberg** [comment]
@Tenischev yo, you might want to look at one new thing I added to this PR, it is a readme template that was initially missing. Have a look at `_README.md`

File: _README.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+- 0.50.0 =< [Generator](https://github.com/asyncapi/generator/) < 2.0.0,
+- Generator specific [requirements](https://github.com/asyncapi/generator/#requirements)
+
+## Specification requirements
+
+<!--  
+The template might need some AsyncAPI properties that normally are optional. For example code generator might require some specific binding information for a given protocol. Even though you can provide defaults or fallbacks, you should describe in the readme what is the most optimal set of properties that the user should provide in the AsyncAPI file.
+-->
+
+Below table contains information about what parts of the specification are required by this template so it generates proper output.
```
Comment (bszwarc) [inline _README.md:44]:
```suggestion
The table contains information on parts of the specification required by this template to generate the proper output.
```

File: _README.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+---|---|---|---
+`components.schemas` | This template supports only schemas that have unique and human-readable names. Such names can also be provided if schemas are described under `components.schemas` and each schema is a separate object with its unique key. | - | -
+
+## Supported protocols
+
+<!--  
+Specify what protocols is your code generator supporting. This information should match the information provided in the template configuration under the `supportedProtocols` property. Don't put this section in your readme if your template doesn't generate code.
+-->
+
+_This template doesn't generate code and therefore is not limited to any specific protocols_
```
Comment (bszwarc) [inline _README.md:56]:
```suggestion
_This template doesn't generate code and therefore is not limited to any specific protocols._
```

File: _README.md [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+## How to use the template
+
+<!--  
+Make sure it is easy to try out the template and check what it generates. Instructions for CLI and Docker should be easy to use; just copy/paste to the terminal. In other words, you should always make sure to have ready to use docker-compose set up so the user can quickly check how generated code behaves.
+-->
+
+This template must be used with the AsyncAPI Generator. All available options on how to generate output are available [here](https://github.com/asyncapi/generator/).
```
Comment (bszwarc) [inline _README.md:64]:
```suggestion
This template must be used with the AsyncAPI Generator. You can find all available options [here](https://github.com/asyncapi/generator/).
```

File: _README.md [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+asyncapi/generator -o /app/output https://raw.githubusercontent.com/asyncapi/generator/v1.0.0-rc.15/test/docs/dummy.yml @asyncapi/markdown-template --force-write
+```
+
+## Template configuration
+
+<!--  
+This information should match the information provided in the template configuration under the `parameters` property.
+-->
+
+You can configure this template using different parameters. You pass them like this in the Generator CLI: `-p PARAM1_NAME=PARAM1_VALUE -p PARAM2_NAME=PARAM2_VALUE`
```
Comment (bszwarc) [inline _README.md:88]:
```suggestion
You can configure this template by passing different parameters in the Generator CLI: `-p PARAM1_NAME=PARAM1_VALUE -p PARAM2_NAME=PARAM2_VALUE`
```
is it only Generator CLI, or there are some other options to pass them and this is only the example?

File: _README.md [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+
+Name | Description | Required | Default | Allowed Values | Example
+---|---|---|---|---|---
+asyncapiFileDir | Custom location of the AsyncAPI file that you provided as an input in a generation. | No | The root of the output directory. | Path to the custom directory relative to the root of output directory provided as a string | `/custom/dir`
```
Comment (bszwarc) [inline _README.md:93]:
```suggestion
asyncapiFileDir | Custom location of the AsyncAPI file that you provided as an input in a generation. | No | The root of the output directory. | Path to the custom directory relative to the root of output directory provided as a string. | `/custom/dir`
```

File: _README.md [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+pdf | Generate PDF from initially generated HTML file | No | `false` | `true`/`false` | `true`
```
Comment (bszwarc) [inline _README.md:94]:
```suggestion
pdf | Generate PDF from initially generated HTML file. | No | `false` | `true`/`false` | `true`
```
If it's a description of a parameter, maybe it is better to use  third person singular, as in `generates` 🤔

File: _README.md [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+png | Generate PNG with a schemas relations diagram. | No | `false` | `true`/`false` | `true`
+svg | Generate SVG with a schemas relations diagram. | No | `false` | `true`/`false` | `true`
+maxTextSize | It is possible that in case of an extremely large AsyncAPI document default mermaid recommended text size will not be enough. Then you need to make it larger explicitly. | No | `50000` |  Number higher than 50000 | `70000`
```
Comment (bszwarc) [inline _README.md:97]:
```suggestion
maxTextSize | It is possible that if an AsyncAPI document is extremely large, the default mermaid recommended text size will not be enough. In such a case, you need to configure it explicitly. | No | `50000` |  A number higher than 50000 | `70000`
```

File: _README.md [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+
+## Custom hooks that you can disable
+
+<!--  
+Document hooks that users can disable and template will still work as expected. Remember that a specific hook can be disabled only if it has a name. In other words, make sure your hook functions are not anonymous.
+-->
+
+Functionality of this template is extended with different hooks that you can disable. You disable them like this in the Generator CLI: `-d HOOK_TYPE1=HOOK_NAME1 -d HOOK_TYPE1=HOOK_NAME2 -d HOOK_TYPE2`
```
Comment (bszwarc) [inline _README.md:105]:
```suggestion
The functionality of this template is extended with different hooks that you can disable like this in the Generator CLI: `-d HOOK_TYPE1=HOOK_NAME1 -d HOOK_TYPE1=HOOK_NAME2 -d HOOK_TYPE2`
```

File: _README.md [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+---|---|---|---|---|---
+generate:after | createAsyncapiFile | It creates AsyncAPI file with content of the spec file passed to the generator
+
+## Development
+
+<!--  
+This section will look the same everywhere, just make sure it references your template.
+-->
+
+The most straightforward command to use this Template is:
```
Comment (bszwarc) [inline _README.md:118]: Is it `Template` or `Template`?  (capitalized vs not capitalized) -> Is `Template` here something different, like an Object? I can see it appears later in the text.

File: _README.md [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+```bash
+ag https://raw.githubusercontent.com/asyncapi/generator/v1.0.0-rc.15/test/docs/dummy.yml https://github.com/asyncapi/template-for-generator-templates -o output
+```
+
+For local development, you need different variations of this command. First of all, you need to know about there important CLI flags:
```
Comment (bszwarc) [inline _README.md:123]:
```suggestion
For local development, you need different variations of this command. First of all, you need to know about three important CLI flags:
```

File: _README.md [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+<!-- tocstop -->
+
+## Overview
+
+<!--  
+The overview should explain in just a few sentences the template's purpose and its most essential features.
+-->
+
+This template generates an HTML page with an overview of the data model provided in your AsyncAPI file. Its core part is the diagram showing relations between the data model. Except for the HTML, you can also generate PDF from the HTML page or get the diagram into a separate PNG and SVG.
```
Comment (bszwarc) [inline _README.md:27]:
```suggestion
This template generates an HTML page with an overview of the data model provided in your AsyncAPI file. Its core part is the diagram showing relations between the data model. Except for HTML, you can also generate PDFs from the HTML page or get the diagram into a separate PNG and SVG.

```

I would remove `the` in front of HTML :) and also perhaps `PDFs` sounds a bit better than `PDF` - if you want to keep it singular, use `a PDF (file)`

File: _README.md [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+
+Name | Description | Required | Default | Allowed Values | Example
+---|---|---|---|---|---
+asyncapiFileDir | Custom location of the AsyncAPI file that you provided as an input in a generation. | No | The root of the output directory. | Path to the custom directory relative to the root of output directory provided as a string | `/custom/dir`
+pdf | Generate PDF from initially generated HTML file | No | `false` | `true`/`false` | `true`
+png | Generate PNG with a schemas relations diagram. | No | `false` | `true`/`false` | `true`
+svg | Generate SVG with a schemas relations diagram. | No | `false` | `true`/`false` | `true`
+maxTextSize | It is possible that in case of an extremely large AsyncAPI document default mermaid recommended text size will not be enough. Then you need to make it larger explicitly. | No | `50000` |  Number higher than 50000 | `70000`
+
+## Custom hooks that you can disable
```
Comment (bszwarc) [inline _README.md:99]: IS `that you can disable` required in the heading? There is some info on disabling in the text already, so I would shorten the heading, unless this info has a specific purpose :)

File: _README.md [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+
+The most straightforward command to use this Template is:
+```bash
+ag https://raw.githubusercontent.com/asyncapi/generator/v1.0.0-rc.15/test/docs/dummy.yml https://github.com/asyncapi/template-for-generator-templates -o output
+```
+
+For local development, you need different variations of this command. First of all, you need to know about there important CLI flags:
+- `--debug` enables the debug mode in Nunjucks engine what makes filters debugging simpler, 
+- `--watch-template` enables a watcher of changes that you make in the Template. It regenerates your Template whenever it detects a change,
+- `--install` enforces reinstallation of the Template
```
Comment (bszwarc) [inline _README.md:126]:
```suggestion
- `--debug` enables the debug mode in Nunjucks engine what makes filters debugging simpler. 
- `--watch-template` enables a watcher of changes that you make in the Template. It regenerates your Template whenever it detects a change.
- `--install` enforces reinstallation of the Template.
```

Either commas at the end of each sentence or no commas at all :)

**Review by bszwarc — COMMENTED** [review]
Added some comments to the readme :)

File: _README.md (hunk 15 above)
Comment (Tenischev) [reply _README.md:99]: I'm fine with current name, because usually user will not even think that he can disable some functionality (in my opinion).

**Comment by Tenischev** [comment]
@derberg, Barbara made a great review, nothing to add 😃

**Review by fmvilas — APPROVED** [review]
Looks good to me ™️ 

https://youtu.be/X_miU0av0qY?t=392

File: _README.md (hunk 12 above)
Comment (derberg) [reply _README.md:119]: yeah, should be `template` as it is something I would compare to words like `repository` or `project`, something specific but not unique, like Generator. Changing it now

File: _README.md (hunk 15 above)
Comment (derberg) [reply _README.md:100]: yeah, it is just purely about disabling, no other reason for a user to read this section, that is why it is in the heading

**Comment by derberg** [comment]
@bszwarc I applied some changes, not sure if you want to check before I merge

**Pull request merged**